lwcomponent/http.go

```diff
@@ -19,58 +19,109 @@
 package lwcomponent
 
 import (
-	"io"
-	"net/http"
+	"fmt"
+	"log"
 	"os"
+	"strconv"
 	"time"
+
+	"github.com/abiosoft/colima/util/terminal"
+	"github.com/go-resty/resty/v2"
 )
 
 const (
-	defaultTimeout  = 30 * time.Second
-	defaultMaxRetry = 2
+	DefaultMaxRetry = 3
 )
 
-func DownloadFile(filepath string, url string, timeout time.Duration) (err error) {
-	if timeout == 0 {
-		timeout = defaultTimeout
-	}
+// Retry 3 times (4 requests total)
+// Resty default RetryWaitTime is 100ms
+// Exponential backoff to a maximum of RetryWaitTime of 2s
+func DownloadFile(filepath string, url string, size int64) (err error) {
+	client := resty.New()
 
-	client := &http.Client{Timeout: timeout}
+	download_timeout := os.Getenv("CDK_DOWNLOAD_TIMEOUT_MINUTES")
+	if download_timeout != "" {
+		val, err := strconv.Atoi(download_timeout)
 
-	file, err := os.Create(filepath)
-	if err != nil {
-		return err
+		if err == nil {
+			client.SetTimeout(time.Duration(val) * time.Minute)
+		}
 	}
-	defer file.Close()
 
-	err = downloadFile(client, url, file)
+	client.SetRetryCount(DefaultMaxRetry)
 
-	for retry := 0; retry < defaultMaxRetry && os.IsTimeout(err); retry++ {
-		_, err = file.Seek(0, io.SeekStart)
-		if err != nil {
-			return err
+	client.OnError(func(req *resty.Request, err error) {
+		if v, ok := err.(*resty.ResponseError); ok {
+			fmt.Println(v.Response.Body())
 		}
+	})
 
-		err = downloadFile(client, url, file)
+	_, err = os.Create(filepath)
+	if err != nil {
+		return
 	}
 
+	done := make(chan int64)
+
+	go downloadProgress(done, filepath, size)
+
+	response, err := client.R().SetOutput(filepath).Get(url)
+
+	done <- response.Size()
+
 	return
 }
 
-// client.Get returns on receiving HTTP headers and we stream the HTTP data to the output file.
-// A timeout will interrupt io.Copy.
-func downloadFile(client *http.Client, url string, file *os.File) (err error) {
+func downloadProgress(done chan int64, filepath string, totalSize int64) {
+	file, err := os.Open(filepath)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer file.Close()
+
+	vterm := terminal.NewVerboseWriter(10)
+	defer vterm.Close()
+
 	var (
-		resp *http.Response
+		previous float64 = 0
+		stop     bool    = false
 	)
 
-	resp, err = client.Get(url)
-	if err != nil {
-		return err
-	}
-	defer resp.Body.Close()
+	for !stop {
+		select {
+		case <-done:
+			stop = true
+		default:
+			info, err := file.Stat()
+			if err != nil {
+				log.Fatal(err)
+			}
 
-	_, err = io.Copy(file, resp.Body)
+			size := info.Size()
+			if size == 0 {
+				size = 1
+			}
 
-	return
+			if totalSize == 0 {
+				mb := float64(size) / (1 << 20)
+
+				if mb > previous {
+					_, _ = vterm.Write([]byte(fmt.Sprintf("Downloaded: %.0fmb\n", mb)))
+
+					previous = mb
+				}
+			} else {
+				percent := float64(size) / float64(totalSize) * 100
+
+				if percent > previous {
+					_, _ = vterm.Write([]byte(fmt.Sprintf("Downloaded: %.0f", percent)))
+					_, _ = vterm.Write([]byte("%"))
```

Comment:
Not familiar with `vterm` but I think the previous commit for this had `println` so there'd be a new line after the percent sign - will there still be a new line?